server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java

```diff
@@ -206,24 +230,55 @@ public ClusterTasksResult<JoinTaskExecutor.Task> execute(ClusterState currentSta
         transportService.registerRequestHandler(
             VALIDATE_JOIN_ACTION_NAME,
             ThreadPool.Names.GENERIC,
-            ValidateJoinRequest::new,
+            BytesTransportRequest::new,
             (request, channel, task) -> {
+                handleValidateJoinRequest(currentStateSupplier, joinValidators, request);
+                channel.sendResponse(Empty.INSTANCE);
+            }
+        );
+    }
+
+    private void handleValidateJoinRequest(
+        Supplier<ClusterState> currentStateSupplier,
+        Collection<BiConsumer<DiscoveryNode, ClusterState>> joinValidators,
+        BytesTransportRequest request
+    ) throws IOException {
+        final Compressor compressor = CompressorFactory.compressor(request.bytes());
+        StreamInput in = request.bytes().streamInput();
+        final ClusterState incomingState;
+        try {
+            if (compressor != null) {
+                in = new InputStreamStreamInput(compressor.threadLocalInputStream(in));
+            }
+            in = new NamedWriteableAwareStreamInput(in, namedWriteableRegistry);
+            in.setVersion(request.version());
+            // If true we received full cluster state - otherwise diffs
+            if (in.readBoolean()) {
+                // Close early to release resources used by the de-compression as early as possible
+                try (StreamInput input = in) {
+                    incomingState = ClusterState.readFrom(input, transportService.getLocalNode());
+                } catch (Exception e) {
+                    logger.warn("unexpected error while deserializing an incoming cluster state", e);
```

Comment:
This would be common code with cluster state publishing deserialization. Can we move this also to common place?